src/refda/exec.c

```diff
@@ -373,93 +375,434 @@ void *refda_exec_worker(void *arg)
+    {
+        CACE_LOG_ERR("Invalid TBR %p action, unable to enable the rule", tbr);
+        return 1;
+    }
+
+    // Adjust rule state
+    tbr->enabled    = true;
+    tbr->exec_count = 0; // Ensure count is reset when rule is enabled
+    atomic_fetch_add(&agent->instr.num_tbrs, 1);
+
+    // Schedule initial rule execution
+    int result = refda_exec_schedule_tbr(agent, tbr, true);
+    return result;
+}
+
+int refda_exec_tbr_disable(refda_agent_t *agent, refda_amm_tbr_desc_t *tbr)
+{
+    CHKERR1(tbr);
+    tbr->enabled = false;
+    atomic_fetch_sub(&agent->instr.num_tbrs, 1);
+    return 0;
+}
+
+static int refda_exec_schedule_sbr(refda_agent_t *agent, refda_amm_sbr_desc_t *sbr);
+
+static int refda_exec_check_sbr_condition(refda_agent_t *agent, refda_amm_sbr_desc_t *sbr, cace_ari_t *result)
+{
+    refda_runctx_t runctx;
+    refda_runctx_init(&runctx);
+
+    if (refda_runctx_from(&runctx, agent, NULL))
+    {
+        return 2;
+    }
+
+    cace_ari_t ari_res = CACE_ARI_INIT_UNDEFINED;
+    int        res     = refda_eval_target(&runctx, &ari_res, &(sbr->condition));
+
+    if (res)
+    {
+        CACE_LOG_ERR("Unable to evaluate SBR condition");
+    }
+    else
+    {
+        const cace_amm_type_t *typeobj = cace_amm_type_get_builtin(CACE_ARI_TYPE_BOOL);
+        res                            = cace_amm_type_convert(typeobj, result, &ari_res);
+        if (res)
+        {
+            CACE_LOG_ERR("Unable to convert SBR condition result to boolean");
+        }
+    }
+
+    cace_ari_deinit(&ari_res);
+    refda_runctx_deinit(&runctx);
+
+    return res;
+}
+
+/** Begin a single execution of a state based rule
+ */
+static void refda_exec_sbr(refda_agent_t *agent, refda_amm_sbr_desc_t *sbr)
+{
+    CHKERR1(agent);
+    CHKERR1(sbr);
+
+    if (!sbr->enabled)
+    {
+        CACE_LOG_INFO("SBR %p is not enabled", sbr);
+        return;
+    }
+
+    if (refda_amm_sbr_desc_reached_max_exec_count(sbr))
+    {
+        CACE_LOG_INFO("SBR %p reached maximum execution count", sbr);
+        refda_exec_sbr_disable(agent, sbr);
+        return;
+    }
+
+    cace_ari_t ari_result = CACE_ARI_INIT_UNDEFINED;
+    int        result     = refda_exec_check_sbr_condition(agent, sbr, &ari_result);
+
+    if (!result)
+    {
+        bool bool_result = false;
+        result           = cace_ari_get_bool(&ari_result, &bool_result);
+        CACE_LOG_INFO("SBR %p condition is %d", sbr, bool_result);
+
+        if (!result && bool_result)
+        {
+            refda_exec_seq_t *seq = refda_exec_seq_list_push_back_new(agent->exec_state);
+            seq->pid              = agent->exec_next_pid++;
+
+            if (!refda_exec_rule_action(agent, seq, &(sbr->action)))
+            {
+                sbr->exec_count++;
+                atomic_fetch_add(&agent->instr.num_sbrs_trig, 1);
             }
         }
+    }
+
+    // Always schedule next execution of the rule
```

Comment:
The schedule-next-evaluation should happen at the start of this function so that the evaluation interval is consistent and independent of the complexity of its condition.

Reply:
Good idea, the scheduling code has been relocated accordingly.